src/signature.c

```diff
@@ -1526,9 +1527,9 @@ static int p11prov_rsasig_set_ctx_params(void *ctx, const OSSL_PARAM params[])
             CK_RV rv;
 
             rv = p11prov_check_mechanism(sigctx->provctx,
-                                         p11prov_obj_get_slotid(sigctx->key),
+                                         slotid,
                                          CKM_RSA_PKCS_PSS);
-            if (rv != CKR_OK) {
+            if (rv != CKR_OK && slotid != CK_UNAVAILABLE_INFORMATION) {
```

Comment:
While I can see why you are doing this I think we should take a different approach.
After line 1487 you should immediately check is slotid is valid, and if not fallback to pull the slotid of the login session which is the session we use to store the key if we ever need to perform an operation.

This way the check on the mechanism will be always done on the correct slot.

Reply:
Use
```
slotid = p11prov_get_default_slot(slots);
```
in the fallback and fail the operation if no default slot is available..

Reply:
Thank you! Yes, this was mostly hackish solution to prove the code path is causing the issue not meant for final solution. Using login/default slot makes the most sense.